lib/SILOptimizer/Analysis/MemoryBehavior.cpp

```diff
@@ -241,10 +241,17 @@ MemBehavior MemoryBehaviorVisitor::visitApplyInst(ApplyInst *AI) {
   MemBehavior Behavior = MemBehavior::None;
 
   // We can ignore mayTrap().
-  if (ApplyEffects.mayReadRC() ||
-      (InspectionMode == RetainObserveKind::ObserveRetains &&
-       ApplyEffects.mayAllocObjects())) {
-    Behavior = MemBehavior::MayHaveSideEffects;
+  if (any_of(enumerate(AI->getArgumentOperands()),
```

Comment:
@eeckstein so I think there's a bug here: it seems that I can't take the indices from `AI->getArgumentOperands()` and use them here `AI->getSubstCalleeType()->getParameters()[op.Index]`.

After this change this showed up whilst compiling the stdlib as 

```
Assertion failed: (Index < Length && "Invalid index!"), function operator[], file /Users/johannes/devel/llvm/include/llvm/ADT/ArrayRef.h, line 233.
0  swift                    0x000000010afecdc8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x000000010afebd26 llvm::sys::RunSignalHandlers() + 86
2  swift                    0x000000010afed38e SignalHandler(int) + 366
3  libsystem_platform.dylib 0x00007fff6e8d8f5a _sigtramp + 26
4  libsystem_platform.dylib 0x0000000100000008 _sigtramp + 2440196296
5  libsystem_c.dylib        0x00007fff6e703cea abort + 127
6  libsystem_c.dylib        0x00007fff6e6cbd40 basename_r + 0
7  swift                    0x00000001086ecf90 swift::SILVisitor<(anonymous namespace)::MemoryBehaviorVisitor, swift::SILInstruction::MemoryBehavior>::visit(swift::ValueBase*) + 7872
```

In other words it seems that there's not a 1:1 match of `AI->getSubstCalleeType()->getParameters()` and `AI->getArgumentOperands()`. Can you confirm that and what's the right way to go from an ApplyInst 'operand' to a parameter?